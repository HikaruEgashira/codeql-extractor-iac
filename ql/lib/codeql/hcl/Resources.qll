private import codeql.Locations
private import codeql.hcl.AST

// Resources are the most important element in the Terraform language.
// Each resource block describes one or more infrastructure objects, such as
// virtual networks, compute instances, or higher-level components such as DNS
// records.
//
// https://developer.hashicorp.com/terraform/language/resources/syntax
class Resource extends Block {
  Resource() { this.hasType("resource") }

  string getName() { result = this.getLabel(1) }

  string getResourceType() { result = this.getLabel(0) }

  override string toString() {
    result = "resource " + this.getResourceType() + " " + this.getName()
  }
}
