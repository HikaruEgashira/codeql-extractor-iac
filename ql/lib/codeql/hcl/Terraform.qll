private import codeql.files.FileSystem

class TerraformFiles extends File {
  TerraformFiles() { this.getBaseName().regexpMatch(".*.(tf|hcl)$") }
}
