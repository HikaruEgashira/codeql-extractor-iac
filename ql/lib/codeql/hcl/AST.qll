import codeql.hcl.ast.AstNodes
import codeql.hcl.ast.Block
import codeql.hcl.ast.Calls
import codeql.hcl.ast.Literals
import codeql.hcl.ast.Variables
import codeql.hcl.ast.Attributes
