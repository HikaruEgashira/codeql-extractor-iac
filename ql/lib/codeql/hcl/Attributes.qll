import codeql.iac.ast.internal.TreeSitter
import codeql.iac.ast.internal.Hcl
import codeql.hcl.AST
import codeql.hcl.ast.AstNodes
import codeql.hcl.Resources

private Expr getAttributeRef(Attribute attr) {
  exists(GetAttrExpr e | e = attr.getExpr() |
    // variable / var
    e.getExpr().(Variable).getName() = "var" and
    exists(Block var |
      var.getType() = "variable" and var.getLabel(0) = e.getKey().(Identifier).getName()
    |
      result = var.getAttribute("default")
    )
    or
    // resource lookup
    exists(Resource resources |
      resources.getResourceType() = e.getExpr().(GetAttrExpr).getExpr().(Variable).getName() and
      resources.getLabel(1) = e.getExpr().(GetAttrExpr).getKey().(Identifier).getName()
    |
      // correct
      result = resources.getAttribute(e.getKey().(Identifier).getName())
    )
  )
}

/**
 * A Terraform attribute.
 */
class Attribute extends Expr, TAttribute {
  private HCL::Attribute attribute;
  private Expr reference;

  Attribute() { this = TAttribute(attribute) or reference = getAttributeRef(this) }

  override string getAPrimaryQlClass() { result = "Attribute" }

  override string toString() {
    result = this.getKey().getName() + " = " + this.getExpr().toString()
  }

  Identifier getKey() { toHclTreeSitter(result) = attribute.getKey() }

  Expr getExpr() { result = reference or toHclTreeSitter(result) = attribute.getVal() }

  Expr getReference() { result = reference }
}
