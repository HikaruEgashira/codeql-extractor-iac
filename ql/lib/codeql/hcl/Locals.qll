import hcl

Expr getLocal(string name) {
  exists(Block locals | locals.getType() = "locals" and result = locals.getAttribute(name))
}
