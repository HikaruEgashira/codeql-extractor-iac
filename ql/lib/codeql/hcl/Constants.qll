import hcl
import Locals

private string evalStringConstant(Expr e) {
  result = e.(StringLiteral).getValue()
  or
  e.(GetAttrExpr).getExpr().(Variable).getName() = "local" and
  result = evalStringConstant(getLocal(e.(GetAttrExpr).getKey().getName()))
  // TODO: handle string interpolation
}

class ConstantExpr extends Expr {
  string value;

  ConstantExpr() { value = evalStringConstant(this) }

  string getStringValue() { result = value }
}
