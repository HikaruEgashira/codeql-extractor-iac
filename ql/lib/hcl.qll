import codeql.Locations
import codeql.files.FileSystem
import codeql.hcl.ast.AstNodes
import codeql.hcl.Resources
import codeql.hcl.Constants
import codeql.hcl.Locals
import codeql.hcl.Terraform
// providers
import codeql.hcl.providers.AWS
import codeql.hcl.providers.Azure
import codeql.hcl.providers.GCP
import codeql.hcl.providers.Helm
// security
import codeql.hcl.security.HardcodedPasswords
