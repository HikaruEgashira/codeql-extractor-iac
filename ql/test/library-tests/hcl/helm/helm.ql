import hcl

query predicate helm(TerraformHelmChart::HelmResource r) { any() }

query predicate provider(TerraformHelmChart::HelmProvider p) { any() }

query predicate release(TerraformHelmChart::Release r) { any() }

query predicate releaseSets(TerraformHelmChart::Release r, TerraformHelmChart::ReleaseSet s) {
  r.getSets() = s
}
