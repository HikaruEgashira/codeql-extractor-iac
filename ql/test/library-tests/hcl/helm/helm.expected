helm
| provider.tf:21:1:31:1 | resource helm_release nginx-ingress-controller |
provider
| provider.tf:1:1:19:1 | provider |
release
| provider.tf:21:1:31:1 | resource helm_release nginx-ingress-controller |
releaseSets
| provider.tf:21:1:31:1 | resource helm_release nginx-ingress-controller | provider.tf:27:3:30:3 | set |
